### PR TITLE
Add Crio support

### PR DESCRIPTION
**Pull request by slabastrou**
See subject. Default is still docker but we can use Crio with the following arguments container-type=crio

### CODE REVIEW

**Comment by kmabda** [comment]
There is no need to add `scripts/add-network-annotation-oc.sh` and create confusion, `scripts/add-network-annotation.sh` should work in ocp

File: controller/crio-runtime/runtime.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+		}
+	}
+
+	namespaces := podStatusResponseInfo.RunTimeSpec.Linux.NameSpaces
+	glog.V(5).Infof("GetNetNS:RunTimeSpec.Linux.NameSpaces:", namespaces)
+	for _, namespace := range namespaces {
+		if namespace.Type == "network" {
+			ss := strings.Split(namespace.Path, "/")
+			netNS := ss[len(ss)-1]
+			glog.V(3).Infof("GetNetNS:NetNS:%s", netNS)
```
[the remark this answers is not in the document]
Comment (slabastrou) [reply controller/crio-runtime/runtime.go:101]: Do you propose lower level glog.Infof or higher V(5) ?

File: controller/cni/cni.go [hunk 2]
```diff
@@ -27,7 +28,11 @@ import (
 	"github.com/containernetworking/cni/libcni"
 	cnitypes "github.com/containernetworking/cni/pkg/types"
 	"github.com/golang/glog"
-	"k8s.io/kubernetes/pkg/kubelet/network"
+	"k8s.io/kubernetes/pkg/kubelet/dockershim/network"
+
+	// "k8s.io/kubernetes/pkg/kubelet/network"
+	// "github.com/containernetworking/cni"
+
```
[the remark this answers is not in the document]
Comment (slabastrou) [reply controller/cni/cni.go]: removed

File: controller/controller.go [hunk 3]
```diff
@@ -29,10 +31,33 @@ import (
...
+	// Crio type container
+	Crio
+)
+
+func (ct ContainerType) String() string {
+	switch ct {
+	case Docker:
+		return "Foo"
+	case Crio:
+		return "Bar"
```
[the remark this answers is not in the document]
Comment (slabastrou) [reply controller/controller.go]: Fixed

File: controller/crio-runtime/runtime.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
+// /*
+// Copyright 2017-2019 Kaloom Inc.
```
[the remark this answers is not in the document]
Comment (slabastrou) [reply controller/crio-runtime/runtime.go]: Fixed

File: controller/crio-runtime/runtime.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	glog.V(5).Infof("GetNetNS:GetInfo():%s", mapInfo)
+	var podStatusResponseInfo PodStatusResponseInfo
+	info := mapInfo["info"]
+	glog.V(5).Infof("GetNetNS:info:%s", info)
+	err = json.Unmarshal([]byte(info), &podStatusResponseInfo)
+	if err != nil {
+		glog.Errorf("GetNetNS:error decoding response:", err)
+		if e, ok := err.(*json.SyntaxError); ok {
+			glog.Errorf("GetNetNS:syntax error at byte offset ", e.Offset)
+		}
```
[the remark this answers is not in the document]
Comment (slabastrou) [reply controller/crio-runtime/runtime.go:92]: Good catch

File: controller/runtime.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
+/*
+Copyright 2017-2019 Kaloom Inc.
```
[the remark this answers is not in the document]
Comment (slabastrou) [reply controller/runtime.go]: done

File: README.md [hunk 7]
```diff
@@ -99,7 +99,8 @@ see [kactus cni-plugin repo](https://github.com/kaloom/kubernetes-kactus-cni-plu
 
 3. delopy the podagent as a daemon set:
 
-> $ `kubectl apply -f manifests/podagent-ds.yaml`
+> $ `kubectl apply -f manifests/podagent-ds.yaml`       for docker
+> $ `kubectl apply -f manifests/podagent-cs.yaml`       for crio
```
[the remark this answers is not in the document]
Comment (slabastrou) [reply README.md:103]: Yeap, added

File: controller/crio-runtime/runtime.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kmabda) [reply controller/crio-runtime/runtime.go:102]: I think V(5) is more appropriate (it's more debugging level)

File: controller/crio-runtime/runtime.go [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	pb "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+	"k8s.io/kubernetes/pkg/kubelet/util"
+)
+
+const (
+	crioNetNSFmt = "/run/netns/%s"
```
Comment (kmabda) [inline controller/crio-runtime/runtime.go:35]: The NS folder is under `/var/run/netns`, although `/var/run` is a symbolic link to `/run` we should not assume it's the case on all system; therefore:

crioNetNSFmt = "/var/run/netns/%s" is more appropriate

File: manifests/podagent-cs.yaml [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        hostPath:
+          path: /var/run/crio.sock
+          type: Socket
+      - name: libcni
+        hostPath:
+          path: /var/lib/cni
+          type: Directory
+      - name: netns
+        hostPath:
+          path: /run/netns
```
Comment (kmabda) [inline manifests/podagent-cs.yaml:84]: path: /var/run/netns

File: manifests/podagent-cs.yaml [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+          mountPath: /opt/cni/bin
+          readOnly: true
+        - name: vendorcnibin
+          mountPath: /opt/kaloom/cni/bin
+          readOnly: true
+        - name: cri
+          mountPath: /var/run/crio.sock
+          readOnly: true
+        - name: libcni
+          mountPath: /var/lib/cni
```
Comment (kmabda) [inline manifests/podagent-cs.yaml:60]: - name: netns
  mountPath: /var/run/netns

File: controller/crio-runtime/runtime.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (slabastrou) [reply controller/crio-runtime/runtime.go]: Ok

File: controller/crio-runtime/runtime.go (hunk 8 above)
Comment (slabastrou) [reply controller/crio-runtime/runtime.go]: Fixed in fps as well

File: manifests/podagent-cs.yaml (hunk 9 above)
Comment (slabastrou) [reply manifests/podagent-cs.yaml]: fixed

File: manifests/podagent-cs.yaml (hunk 10 above)
Comment (slabastrou) [reply manifests/podagent-cs.yaml:60]: fixed

**Pull request merged**